src/libs/safe-path/src/safe_dir_builder.rs

```diff
@@ -0,0 +1,162 @@
+    /// also non-recursive.
+    pub fn new() -> Self {
+        SafeDirBuilder {
+            root: PathBuf::from("/"),
+            mode: 0o777,
+            recursive: false,
+        }
+    }
+
+    /// Indicates that directories should be created recursively, creating all parent directories.
+    /// Parents that do not exist are created with the same security and permissions settings.
+    pub fn recursive(&mut self, recursive: bool) -> &mut Self {
+        self.recursive = recursive;
+        self
+    }
+
+    /// Sets the mode to create new directories with. This option defaults to 0o777.
+    pub fn mode(&mut self, mode: u32) -> &mut Self {
+        self.mode = mode;
+        self
+    }
+
+    /// Sets the root directory for security check.
+    ///
+    /// # Safety
+    /// The `root` is the trusted part, and should not be modifiable by untrusted agents.
+    pub fn root<P: AsRef<Path>>(&mut self, root: P) -> Result<&mut Self> {
+        let root = root.as_ref().canonicalize()?;
+        if !root.is_dir() {
+            return Err(Error::new(
+                ErrorKind::Other,
+                format!("Invalid path: {}", root.display()),
+            ));
+        }
+        self.root = root;
+
+        Ok(self)
+    }
+
+    /// Creates the specified directory with the options configured in this builder.
+    ///
+    /// The `path` must be a subdirectory of `SafePathBuf::root()`, otherwise error will be returned.
+    /// It is considered an error if the directory already exists unless recursive mode is enabled.
+    pub fn create<P: AsRef<Path>>(&self, path: P) -> Result<SafePathBuf> {
+        let mut root = self.root.canonicalize()?;
+        if !root.is_dir() {
+            return Err(Error::new(
+                ErrorKind::Other,
+                format!("Invalid path: {}", root.display()),
+            ));
+        }
+
+        let path = safe_join("/", path)?;
+        let mut suffix = path.strip_prefix(&root).map_err(|_| {
+            Error::new(
+                ErrorKind::Other,
+                format!("Invalid path: {}", path.display()),
+            )
+        })?;
+        if suffix.file_name().is_none() {
+            return SafePathBuf::from_path(root);
+        }
+        if !self.recursive {
+            if let Some(parent) = path.parent() {
+                root = root.join(parent);
+            }
+            // Safe to unwrap() because we have verified `suffix` is not empty.
+            suffix = Path::new(suffix.file_name().unwrap());
+        }
+
+        for comp in suffix {
+            let file = SafePathBuf::from_path(&root)?;
+            if !file.target().is_dir() {
+                return Err(Error::new(
+                    ErrorKind::Other,
+                    format!("Invalid path: {}", root.display()),
+                ));
+            }
+            root = root.join(comp);
+            DirBuilder::new()
+                .mode(self.mode)
+                .recursive(true)
```

Comment:
You're calling this for each component, so you shouldn't need recursive on the underlying `DirBuilder`.

Reply:
good catch!